### PR TITLE
arith between types

### PR DESCRIPTION
**Pull request by 3pmile**

### CODE REVIEW

File: src/integer/z/arithmetic/add.rs [hunk 1]
```diff
@@ -46,8 +47,101 @@ impl Add for &Z {
...
+#[cfg(test)]
+mod test_add_between_types {
+
+    use crate::integer::Z;
+    use std::str::FromStr;
+
+    /// testing addition between different types
+    #[test]
+    fn add() {
+        let a: Z = Z::from_str("42").unwrap();
```
Comment (Marvin-Beckmann) [inline src/integer/z/arithmetic/add.rs:63]: Please also do this for the others, just my personal preference
```suggestion
        let a: Z = Z::from("42");
```

File: src/macros/arithmetics.rs [hunk 2]
```diff
@@ -16,19 +16,22 @@
...
-            type Output = $type;
+    ($trait:ident, $trait_function:ident, $type:ident, $other_type:ident, $output_type:ident) => {
+        #[doc(hidden)]
+        impl $trait<$other_type> for $type {
+            type Output = $output_type;
 
             paste::paste! {
-                #[doc = "Documentation at [`" $type "::" $trait_function "`]."]
-                fn $trait_function(self, other: Self) -> Self::Output {
+                #[doc = "Documentation at [`" $output_type "::" $trait_function "`]."]
```
Comment (Marvin-Beckmann) [inline src/macros/arithmetics.rs:33]: Shouldn't this be $type?
```suggestion
                #[doc = "Documentation at [`" $type "::" $trait_function "`]."]
```

File: src/macros/arithmetics.rs [hunk 3]
```diff
@@ -46,30 +49,35 @@ pub(crate) use arithmetic_trait_borrowed_to_owned;
...
-            type Output = $type;
+    ($trait:ident, $trait_function:ident, $type:ident, $other_type:ident, $output_type:ident) => {
+        #[doc(hidden)]
+        impl $trait<$other_type> for &$type {
+            type Output = $output_type;
             paste::paste! {
-                #[doc = "Documentation at [`" $type "::" $trait_function "`]."]
-                fn $trait_function(self, other: $type) -> Self::Output {
+
+                #[doc = "Documentation at [`" $output_type "::" $trait_function "`]."]
```
Comment (Marvin-Beckmann) [inline src/macros/arithmetics.rs:68]: see above
```suggestion
                #[doc = "Documentation at [`" $type"::" $trait_function "`]."]
```

File: src/macros/arithmetics.rs [hunk 4]
```diff
@@ -46,30 +49,35 @@ pub(crate) use arithmetic_trait_borrowed_to_owned;
...
 
-        impl $trait<&$type> for $type {
-            type Output = $type;
+        #[doc(hidden)]
+        impl $trait<&$other_type> for $type {
+            type Output = $output_type;
             paste::paste! {
-                #[doc = "Documentation at [`" $type "::" $trait_function "`]."]
-                fn $trait_function(self, other: &Self) -> Self::Output {
+                #[doc = "Documentation at [`" $output_type "::" $trait_function "`]."]
```
Comment (Marvin-Beckmann) [inline src/macros/arithmetics.rs:79]:
```suggestion
                #[doc = "Documentation at [`" $type "::" $trait_function "`]."]
```

File: src/integer/z/arithmetic/add.rs [hunk 5]
```diff
@@ -46,8 +47,101 @@ impl Add for &Z {
...
+arithmetic_between_types!(Add, add, Z, i64 i32 i16 i8 u64 u32 u16 u8);
+
+#[cfg(test)]
+mod test_add_between_types {
+
+    use crate::integer::Z;
+    use std::str::FromStr;
+
+    /// testing addition between different types
+    #[test]
```
Comment (Marvin-Beckmann) [inline src/integer/z/arithmetic/add.rs:61]:
```suggestion
    #[test]
    #[allow(clippy::op_ref)]
```

File: src/integer/z/arithmetic/mul.rs [hunk 6]
```diff
@@ -46,8 +47,101 @@ impl Mul for &Z {
...
+arithmetic_between_types!(Mul, mul, Z, i64 i32 i16 i8 u64 u32 u16 u8);
+
+#[cfg(test)]
+mod test_mul_between_types {
+
+    use crate::integer::Z;
+    use std::str::FromStr;
+
+    /// testing multiplication between different types
+    #[test]
```
Comment (Marvin-Beckmann) [inline src/integer/z/arithmetic/mul.rs:61]:
```suggestion
    #[test]
    #[allow(clippy::op_ref)]
```

File: src/integer/z/arithmetic/sub.rs [hunk 7]
```diff
@@ -46,8 +47,101 @@ impl Sub for &Z {
...
+arithmetic_between_types!(Sub, sub, Z, i64 i32 i16 i8 u64 u32 u16 u8);
+
+#[cfg(test)]
+mod test_sub_between_types {
+
+    use crate::integer::Z;
+    use std::str::FromStr;
+
+    /// testing subtraction between different types
+    #[test]
```
Comment (Marvin-Beckmann) [inline src/integer/z/arithmetic/sub.rs:61]:
```suggestion
    #[test]
    #[allow(clippy::op_ref)]
```

**Review by Marvin-Beckmann — CHANGES_REQUESTED** [review]
Looks good, just a few things

File: src/macros/arithmetics.rs (hunk 2 above)
Comment (3pmile) [reply src/macros/arithmetics.rs:33]: no, the documentation should be shown at the output type (the suggestion would link Z+Q to Z and Q+Z to Q, which is unintuitive), maybe an update of the documentation with new examples is needed

Comment (3pmile) [reply src/macros/arithmetics.rs:33]: should be a todo for Q and Zq

**Comment by 3pmile** [comment]
rebase not working, new PR opened

**Pull request closed without merging**